consensus/src/detail/mod.rs

```diff
@@ -67,6 +75,17 @@ pub enum BlockSource {
 }
 
 impl Consensus {
+    pub fn wait_for_all_events(&self) {
+        while self.events_in_progress.load(std::sync::atomic::Ordering::Acquire) > 0 {
+            std::thread::yield_now();
+            std::thread::sleep(Duration::from_millis(10));
+        }
```

Comment:
Would it be useful to have a limit on the number of iterations here and hard exit when some sort of upper bound is reached?

Reply:
Also does `sleep` not yield implicitly?

Reply:
> Would it be useful to have a limit on the number of iterations here and hard exit when some sort of upper bound is reached?

Probably not. This should never fail if we know what we're doing.

> Also does sleep not yield implicitly?

It may or may not yield depending on the OS. Usually it does, but I don't see the harm.

Reply:
> Also it would be nicer if the events_in_progress counter was abstracted in its own wrapper (say, EventWaitBlocker, better name needed) type and moved to util alongside CountTracker and do the waiting inside the destructor of EventWaitBlocker. Consensus would wait for events just by having an instance of EventWaitBlocker as a member. That way the mechanism could be made reusable

Done. I create a struct that blocks automatically on drop. It's now in utils.